### PR TITLE
[SPARK-32079][PYTHON] Remove namedtuple hack by replacing built-in pickle to cloudpickle

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to replace Python's built-in CPickle to CPickle-based cloudpickle (requires Python 3.8+).
For Python 3.7 and below, it still uses the legacy built-in CPickle for the performance matter.

I did a bit of benchmark with basic cases, and I have seen no performance penalty (attached one of the benchmarks below).

### Why are the changes needed?

To remove named tuple hack for the issues such as: SPARK-32079,  SPARK-22674 and SPARK-27810.

### Does this PR introduce _any_ user-facing change?

No.

#### Micro benchmark:

```python
import time
import pickle
from pyspark import cloudpickle

def measure(f):
    start = time.time()
    f()
    end = time.time()
    print(end - start)

data = [123, "abc", (1, 2, 3), 2.2] * 100000000
measure(lambda: pickle.dumps(data))
measure(lambda: cloudpickle.dumps(data))
measure(lambda: pickle.loads(pickle.dumps(data)))
measure(lambda: cloudpickle.loads(cloudpickle.dumps(data)))
```

```
5.1765618324279785
5.2591071128845215
12.457043886184692
12.1910879611969
```

```python
import time
import random
import pickle
from pyspark import cloudpickle

def measure(f):
    start = time.time()
    f()
    end = time.time()
    print(end - start)

rand_data = []

for _ in range(10000000):
    data = [
        random.randint(1, 100),
        str(random.randint(1, 100)),
        (random.randint(1, 100), random.randint(2, 200), random.randint(3, 300)),
        random.random()
    ]
    random.shuffle(data)
    rand_data.append(data)

measure(lambda: pickle.dumps(rand_data))
measure(lambda: cloudpickle.dumps(rand_data))
measure(lambda: pickle.loads(pickle.dumps(rand_data)))
measure(lambda: cloudpickle.loads(cloudpickle.dumps(rand_data)))
```

```
7.736639976501465
7.8458099365234375
20.306012868881226
17.787282943725586
```

#### E2E benchmark:

```bash
./bin/pyspark --conf spark.python.profile=true
```

```python
import time
from collections import namedtuple
rdd = sc.parallelize([123] * 30000000)
rdd.count()  # init
start = time.time()
rdd.map(lambda x: x).count()
print(time.time() - start)
sc.show_profiles()
```

Before:

2.3216118812561035 (sec)

```
============================================================
Profile of RDD<id=2>
============================================================
         60264297 function calls (60264265 primitive calls) in 22.309 seconds

   Ordered by: internal time, cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 30000016   13.127    0.000   29.890    0.000 rdd.py:1291(<genexpr>)
       32    4.559    0.142   34.449    1.077 {built-in method builtins.sum}
 30000000    3.723    0.000    3.723    0.000 <stdin>:1(<lambda>)
    29297    0.699    0.000    0.699    0.000 {built-in method _pickle.loads}
    29313    0.059    0.000    0.874    0.000 serializers.py:151(_read_with_length)
    58610    0.045    0.000    0.045    0.000 {method 'read' of '_io.BufferedReader' objects}
    29313    0.035    0.000    0.057    0.000 serializers.py:567(read_int)
    29313    0.025    0.000    0.899    0.000 serializers.py:135(load_stream)
    29297    0.016    0.000    0.715    0.000 serializers.py:435(loads)
    29313    0.013    0.000    0.013    0.000 {built-in method _struct.unpack}
    29329    0.006    0.000    0.006    0.000 {built-in method builtins.len}
       16    0.000    0.000    0.000    0.000 rdd.py:409(func)
       16    0.000    0.000    0.001    0.000 serializers.py:256(dump_stream)
...
```

After:

2.279919147491455 (sec)

```
============================================================
Profile of RDD<id=2>
============================================================
         90264361 function calls (90264329 primitive calls) in 34.573 seconds

   Ordered by: internal time, cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 30000016   13.204    0.000   29.982    0.000 rdd.py:1291(<genexpr>)
 30000000   12.087    0.000   15.879    0.000 util.py:77(wrapper)
       32    4.588    0.143   34.571    1.080 {built-in method builtins.sum}
 30000000    3.792    0.000    3.792    0.000 <stdin>:1(<lambda>)
    29297    0.694    0.000    0.694    0.000 {built-in method _pickle.loads}
    29313    0.061    0.000    0.873    0.000 serializers.py:157(_read_with_length)
    58610    0.045    0.000    0.045    0.000 {method 'read' of '_io.BufferedReader' objects}
    29313    0.036    0.000    0.059    0.000 serializers.py:585(read_int)
    29313    0.026    0.000    0.900    0.000 serializers.py:141(load_stream)
    29297    0.018    0.000    0.712    0.000 serializers.py:463(loads)
    29313    0.013    0.000    0.013    0.000 {built-in method _struct.unpack}
    29329    0.007    0.000    0.007    0.000 {built-in method builtins.len}
       16    0.000    0.000   34.573    2.161 worker.py:665(process)
       16    0.000    0.000    0.000    0.000 rdd.py:409(func)
       16    0.000    0.000    0.001    0.000 serializers.py:262(dump_stream)
       16    0.000    0.000    0.001    0.000 cloudpickle_fast.py:59(dumps)
...
```

Existing test cases should cover all test cases.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @ueshin @viirya @BryanCutler @JoshRosen @WeichenXu123 @mengxr FYI

File: python/pyspark/__init__.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def wrapper(self, *args, **kwargs):
     "Accumulator",
     "AccumulatorParam",
     "MarshalSerializer",
-    "PickleSerializer",
+    "CPickleSerializer",
```
Comment (HyukjinKwon) [inline python/pyspark/__init__.py:139]:
```suggestion
    "PickleSerializer",
    "CPickleSerializer",
```

File: python/pyspark/__init__.pyi [hunk 2]
```diff
@@ -38,7 +38,7 @@ from pyspark.profiler import (  # noqa: F401
 from pyspark.rdd import RDD as RDD, RDDBarrier as RDDBarrier  # noqa: F401
 from pyspark.serializers import (  # noqa: F401
     MarshalSerializer as MarshalSerializer,
-    PickleSerializer as PickleSerializer,
+    CPickleSerializer as CPickleSerializer,
```
Comment (HyukjinKwon) [inline python/pyspark/__init__.pyi:41]:
```suggestion
    PickleSerializer as PickleSerializer
    CPickleSerializer as CPickleSerializer,,
```

File: python/pyspark/serializers.py [hunk 3]
```diff
@@ -19,7 +19,7 @@
 PySpark supports custom serializers for transferring data; this can improve
 performance.
 
-By default, PySpark uses :class:`PickleSerializer` to serialize objects using Python's
+By default, PySpark uses :class:`CloudPickleSerializer` to serialize objects using Python's
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py:22]:
```suggestion
By default, PySpark uses :class:`CPickleSerializer` to serialize objects using Python's
```

**Comment by HyukjinKwon** [comment]
I added a couple of micro benchamrks.

File: python/pyspark/serializers.py [hunk 4]
```diff
@@ -19,7 +19,7 @@
 PySpark supports custom serializers for transferring data; this can improve
 performance.
 
-By default, PySpark uses :class:`PickleSerializer` to serialize objects using Python's
+By default, PySpark uses :class:`CPickleSerializer` to serialize objects using Python's
 `cPickle` serializer, which can serialize nearly any Python object.
```
Comment (viirya) [inline python/pyspark/serializers.py:23]: "... using Python's`cPickle` serializer,' <- do we need to update this?

**Review by viirya — COMMENTED** [review]
Have we kept same test coverage for two serializers?

**Review by viirya — COMMENTED** [review]
Looks good if we've not missed current test coverage.

File: python/pyspark/serializers.py (hunk 4 above)
Comment (HyukjinKwon) [inline python/pyspark/serializers.py:23]:
```suggestion
C-based pickle serializer, which can serialize nearly any Python object.
```

**Comment by HyukjinKwon** [comment]
Rebased.

**Comment by HyukjinKwon** [comment]
I'll merge this one in few days if there aren't any comment.

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Thanks, @viirya.

Merged to master.

**Pull request closed without merging**